metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -927,6 +981,31 @@ private void appendRaftEvent(String name, Runnable runnable) {
         }
     }
 
+    /**
+     * A callback for changes to feature levels including metadata.version. This is called synchronously from
+     * {@link FeatureControlManager#replay(FeatureLevelRecord)} which is part of a ControllerWriteEvent. It is safe
+     * to modify controller state here. By the time this listener is called, a FeatureLevelRecord has been committed and
+     * the in-memory state of FeatureControlManager has been updated.
+     */
+    class QuorumFeatureListener implements FeatureLevelListener {
+        @Override
+        public void handle(String featureName, short finalizedVersion) {
+            boolean isActiveController = curClaimEpoch != -1;
+            boolean isFeatureSupported = featureControl.canSupportVersion(featureName, finalizedVersion);
+            if (featureName.equals(MetadataVersion.FEATURE_NAME)) {
+                if (!isFeatureSupported) {
+                    if (isActiveController) {
+                        log.error("Active controller cannot support metadata.version {}", finalizedVersion);
+                    } else {
+                        log.error("Standby controller cannot support metadata.version {}, shutting down.", finalizedVersion);
+                    }
+                    beginShutdown();
+                }
+            }
```

Comment:
Is the behavior dependent on the feature being metadata.version?
```suggestion
            if (!isFeatureSupported) {
                if (isActiveController) {
                    log.error("Active controller cannot support {} {}", featureName, finalizedVersion);
                } else {
                    log.error("Standby controller cannot support {} {}, shutting down.", featureName, finalizedVersion);
                }
                beginShutdown();
            }
```

Reply:
Yea, that's a good point. If the controller receives a record for a feature it can't support it should probably shutdown.